src/psbt.c

```diff
@@ -0,0 +1,2547 @@
+}
+
+
+static void clear_and_free(void *p, size_t len)
+{
+    if (p) {
+        wally_clear(p, len);
+        wally_free(p);
+    }
+}
+
+static bool pubkey_is_compressed(unsigned char pubkey[65]) {
+    return pubkey[0] == 0x02 || pubkey[0] == 0x03;
+}
+
+
+int keypath_map_init_alloc(size_t alloc_len, struct keypath_map **output)
+{
+    struct keypath_map *result;
+
+    TX_CHECK_OUTPUT;
+    TX_OUTPUT_ALLOC(struct keypath_map);
+
+    if (alloc_len) {
+        result->items = wally_malloc(alloc_len * sizeof(*result->items));
+        if (!result->items) {
+            wally_free(result);
+            *output = NULL;
+            return WALLY_ENOMEM;
+        }
+        wally_clear(result->items, alloc_len * sizeof(*result->items));
+    }
+    result->items_allocation_len = alloc_len;
+    result->num_items = 0;
+    return WALLY_OK;
+}
+
+int keypath_map_free(struct keypath_map *keypaths)
+{
+    size_t i;
+
+    if (keypaths) {
+        for (i = 0; i < keypaths->num_items; ++i) {
+            if (keypaths->items[i].origin.path) {
+                wally_clear(keypaths->items[i].origin.path, keypaths->items[i].origin.path_len);
+            }
+        }
+        clear_and_free(keypaths->items, keypaths->num_items * sizeof(*keypaths->items));
+        wally_clear(keypaths, sizeof(*keypaths));
+        wally_free(keypaths);
+    }
+    return WALLY_OK;
+}
+
+static struct keypath_map *clone_keypath_map(const struct keypath_map *keypaths)
+{
+    struct keypath_map *result;
+    size_t i;
+
+    if (keypath_map_init_alloc(keypaths->items_allocation_len, &result) != WALLY_OK) {
+        return NULL;
+    }
+
+    for (i = 0; i < keypaths->num_items; ++i) {
+        memcpy(&result->items[i].pubkey, keypaths->items[i].pubkey, 65);
+        memcpy(&result->items[i].origin.fingerprint, keypaths->items[i].origin.fingerprint, 4);
+        if (keypaths->items[i].origin.path) {
+            if (!clone_bytes((unsigned char **)&result->items[i].origin.path, (unsigned char *)keypaths->items[i].origin.path, keypaths->items[i].origin.path_len * sizeof(*keypaths->items[i].origin.path))) {
+                goto fail;
+            }
+            result->items[i].origin.path_len = keypaths->items[i].origin.path_len;
+        }
+    }
+    result->num_items = keypaths->num_items;
+
+    return result;
+
+fail:
+    keypath_map_free(result);
+    return NULL;
+}
+
+int add_new_keypath(struct keypath_map *keypaths,
+                    unsigned char pubkey[65],
+                    unsigned char fingerprint[4],
+                    uint32_t *path,
+                    size_t path_len)
+{
+    size_t latest;
+
+    if (keypaths->num_items == keypaths->items_allocation_len) {
+        size_t orig_alloc_len = keypaths->items_allocation_len;
+        if (keypaths->items_allocation_len == 0) {
+            keypaths->items_allocation_len = 1;
+        }
+        keypaths->items_allocation_len *= 2;
+        struct keypath_map *new_keypaths = clone_keypath_map(keypaths);
```

Comment:
This feels to me like an unnecessary memory allocation, couldn't it just allocate a new items array without allocating a new keypath_map?